space_packet_parser/definitions.py

```diff
@@ -385,269 +359,326 @@ def _get_container_base_container(
+                result = read_bytes_from_source(binary_data, read_size_bytes=buffer_read_size_bytes)
+                if not result:  # If there is verifiably no more data to add, break
+                    break
+                read_buffer += result
+            # Skip the header bytes
+            current_pos += skip_header_bytes
+            header_bytes = read_buffer[current_pos:current_pos + CCSDS_HEADER_LENGTH_BYTES]
+            header = self._parse_header(header_bytes)
+
+            # per the CCSDS spec
+            # 4.1.3.5.3 The length count C shall be expressed as:
+            #   C = (Total Number of Octets in the Packet Data Field) – 1
+            n_bytes_data = header['PKT_LEN'].raw_value + 1
+            n_bytes_packet = CCSDS_HEADER_LENGTH_BYTES + n_bytes_data
+
+            # Consider it a counted packet once we've parsed the header
+            # and update the number of bits parsed
+            n_packets_parsed += 1
+            n_bytes_parsed += skip_header_bytes + n_bytes_packet
+            if ccsds_headers_only is True:
+                # update the current position to the end of the packet data
+                current_pos += n_bytes_packet
+                p = parseables.CCSDSPacket(raw_data=read_buffer[current_pos - n_bytes_packet:current_pos], **header)
+                yield p
+                continue
+
+            # Based on PKT_LEN fill buffer enough to read a full packet
+            while len(read_buffer) - current_pos < n_bytes_packet:
+                result = read_bytes_from_source(binary_data, read_size_bytes=buffer_read_size_bytes)
+                if not result:  # If there is verifiably no more data to add, break
+                    break
+                read_buffer += result
+
+            # current_pos is still before the header, so we are reading the entire packet here
+            packet_bytes = read_buffer[current_pos:current_pos + n_bytes_packet]
+            current_pos += n_bytes_packet
+            # Wrap the bytes in a class that can keep track of position as we read from it
+            packet = parseables.CCSDSPacket(raw_data=packet_bytes)
+            try:
+                packet = self.parse_CCSDSPacket(packet,
+                                                self.named_containers,
+                                                root_container_name=root_container_name,
+                                                word_size=self.word_size)
+            except UnrecognizedPacketTypeError as e:
+                logger.debug(f"Unrecognized error on packet with APID {header['PKT_APID'].raw_value}'")
+                if yield_unrecognized_packet_errors is True:
+                    # Yield the caught exception without raising it (raising ends generator)
+                    yield e
+                # Continue to next packet
+                continue
+
+            if packet.header['PKT_LEN'].raw_value != header['PKT_LEN'].raw_value:
```

Comment:
I was actually wondering about this check and whether we could avoid reading the first 7 items in the container definition and rely on the CCSDS header being correct. That way we could start the Packet class right away when we are doing the header above and then continue on here. Currently we are parsing the header twice essentially.

Could we check this condition in the definition itself somehow so that when someone is creating the definition from the XTCE we would raise immediately before parsing even began?

Reply:
I think this is probably possible and would be another nice performance enhancement but it has enough impact that I don't really want to tackle that on this PR. 

I think it would basically require us to be able to split the calls to `parse_CCSDSPacket` into two separate calls -- one to parse the header and a second to continue the parsing for the rest of the packet. We would need to rely on field ordering in order to pull out the necessary packet length field. 

Also, you could envision a pathological case where a user has decided they don't really care about the header information so they just define the header as a single blob of 6 bytes rather than splitting it into distinct fields, which would be valid XTCE but would break the ability to grab the packet length field out of the user-defined header representation.

Reply:
Yes, sounds good to leave this for the future.

> Also, you could envision a pathological case where a user has decided they don't really care about the header information so they just define the header as a single blob of 6 bytes rather than splitting it into distinct fields, which would be valid XTCE but would break the ability to grab the packet length field out of the user-defined header representation.

I think we would fail there because we are reading in PKT_LEN bytes to create the "raw bytes to be parsed" object to pass to our XTCE definition parsing (i.e. one full packet's worth of bytes as defined by the header and no more/less). So there has to be at least a valid APID and PKT_LEN attribute in the header I think.

Reply:
💯 But if we require header items to be present as predictable, separate items, we are levying requirements on the XTCE that aren't in the spec. I think it would be completely valid for a user parsing a simple packet to just lump all the header items into a 6 byte binary blob and ignore it. I don't think that's common but I don't think it's any kind of violation to say "I don't care about the headers. My packets are fixed length and I only want the user data."

Maybe it's OK to levy those additional requirements for the library. What do you think?